compiler/lib/effects.ml

```diff
@@ -0,0 +1,552 @@
+end = struct
+  type cont = Var.t
+
+  type continuation = Var.t
+
+  type stack = Var.t
+
+  let pop_fiber k =
+    let res = Var.fresh () in
+    let stack = Var.fresh_n "stack" in
+    let k' = Var.fresh_n "cont" in
+    ( [ Let (res, Prim (Extern "caml_pop_fiber", [ Pv k ]))
+      ; Let (stack, Field (res, 0))
+      ; Let (k', Field (res, 1))
+      ]
+    , stack
+    , k' )
+
+  let resume_stack stack k =
+    let k' = Var.fresh_n "cont" in
+    [ Let (k', Prim (Extern "caml_resume_stack", [ Pv stack; Pv k ])) ], k'
+
+  let cons_fiber fiber continuation =
+    (* Assumes a one-level stack (the last component of [fiber] is empty). *)
+    (* let Cons (k, exn_stack, handlers, Empty) = fiber in
+       continuation := Cons (k, exn_stack, handlers, !continuation) *)
+    let stack = Var.fresh () in
+    let k = Var.fresh () and exn_stack = Var.fresh () and b = Var.fresh () in
+    let handlers = Var.fresh () in
+    [ Let (stack, Field (continuation, 0))
+    ; Let (k, Field (fiber, 0))
+    ; Let (exn_stack, Field (fiber, 1))
+    ; Let (handlers, Field (fiber, 2))
+    ; Let (b, Block (0, [| k; exn_stack; handlers; stack |], NotArray))
+    ; Set_field (continuation, 0, b)
+    ]
+
+  let get_effect_handler stack =
+    (* let Cons (_, _, h, _) = stack in h.effc *)
+    let a = Var.fresh () in
+    let handler = Var.fresh_n "effect_handler" in
+    [ Let (a, Field (stack, 2)); Let (handler, Field (a, 2)) ], handler
+
+  let continuation_of_stack stack =
+    let continuation = Var.fresh_n "continuation" in
+    [ Let (continuation, Block (245, [| stack |], NotArray)) ], continuation
+
+  let push_trap exn_handler =
+    [ Let (Var.fresh (), Prim (Extern "caml_push_trap", [ Pv exn_handler ])) ]
+
+  let pop_trap () =
+    let exn_handler = Var.fresh_n "raise" in
+    [ Let (exn_handler, Prim (Extern "caml_pop_trap", [])) ], exn_handler
+end
+
+let cps_branch ~st (pc, args) =
+  let ret = Var.fresh () in
+  [ Let (ret, Apply { f = closure_of_pc ~st pc; args; exact = true }) ], Return ret
+
+let cps_jump_cont ~st cont =
+  let call_block =
+    let body, branch = cps_branch ~st cont in
+    add_block st { params = []; body; branch }
+  in
+  call_block, []
+
+let cps_last ~st (last : last) ~(k : Runtime.cont) : instr list * last =
+  match last with
+  | Return x ->
+      let ret = Var.fresh () in
+      [ Let (ret, Apply { f = k; args = [ x ]; exact = true }) ], Return ret
+  | Raise (x, _) ->
+      let pop_instrs, exn_handler = Runtime.pop_trap () in
+      let ret = Var.fresh () in
+      ( pop_instrs @ [ Let (ret, Apply { f = exn_handler; args = [ x ]; exact = true }) ]
+      , Return ret )
+  | Stop -> [], Stop
+  | Branch cont -> cps_branch ~st cont
+  | Cond (x, cont1, cont2) ->
+      [], Cond (x, cps_jump_cont ~st cont1, cps_jump_cont ~st cont2)
+  | Switch (x, c1, c2) ->
+      (* To avoid code duplication during JavaScript generation, we need
+         to create a single block per continuation *)
```

Comment:
Shouldn't this apply to `Cond` as well ?

Reply:
`cont` are probably always different with `Cond`. Maybe add a comment or an assert ?

Reply:
They are always different. I have added an `assert`.